File: test/ModularPipelines.UnitTests/Execution/AsyncDisposableModuleTests.cs
using ModularPipelines.Context;
using ModularPipelines.Modules;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Execution;

public class AsyncDisposableModuleTests
{
    [Test]
    public async Task SuccessfullyDisposed()
    {
        var pipelineSummary = await TestPipelineHostBuilder.Create()
            .AddModule<AsyncDisposableModule>()
            .ExecutePipelineAsync();
        await Assert.That(pipelineSummary.Modules.OfType<AsyncDisposableModule>().Single().IsDisposed).IsTrue();
    }

    public class AsyncDisposableModule : Module<bool>, IAsyncDisposable
    {
        public bool IsDisposed { get; private set; }

        /// <inheritdoc/>
        public override async Task<bool> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            // Reduced delay from 100ms to 1ms for faster test execution
            await Task.Delay(1, cancellationToken);
            return true;
        }

        /// <inheritdoc/>
        public async ValueTask DisposeAsync()
        {
            // Reduced delay from 100ms to 1ms for faster test execution
            await Task.Delay(1);
            IsDisposed = true;
            GC.SuppressFinalize(this);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Execution/AsyncDisposableModuleTests.fs
namespace ModularPipelines.UnitTests.FSharp.Execution

open System
open System.Threading
open System.Threading.Tasks
open System.Linq
open ModularPipelines.Context
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core
open TUnit.Assertions.Extensions

type private AsyncDisposableModule() =
    inherit Module<bool>()

    member val IsDisposed = false with get, set

    override _.ExecuteAsync(_: IModuleContext, cancellationToken: CancellationToken) =
        task {
            do! Task.Delay(1, cancellationToken)
            return true
        }

    interface IAsyncDisposable with
        member this.DisposeAsync() =
            ValueTask(
                task {
                    do! Task.Delay(1)
                    this.IsDisposed <- true
                    GC.SuppressFinalize(this)
                }
            )

type AsyncDisposableModuleTests() =
    [<Test>]
    member _.SuccessfullyDisposed() = async {
        let! pipelineSummary =
            TestPipelineHostBuilder.Create().AddModule<AsyncDisposableModule>().ExecutePipelineAsync()
            |> Async.AwaitTask

        let isDisposed = pipelineSummary.Modules.OfType<AsyncDisposableModule>().Single().IsDisposed
        do! check(Assert.That(isDisposed).IsTrue())
    }
